<h1>DATA ANALYTICS OF UK ROAD ACCIDENT</h1>

<h3>Analysis: Lovely Claire Cabana</h3>

<h2>1. Importing Necessary Libraries</h2>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

<h2>2. Coverting data set to data frame</h2>

In [31]:
accidents = pd.read_csv('datasets\\uk_road_accident.csv')

<h2>3. Check Dataframe Information</h2>

In [32]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h2>4. Check Discriptive Statistics</h2>

In [33]:
accidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [34]:
accidents.describe().T

count       mean       std        min        25%  \
Latitude              660654.0  52.553866  1.406922  49.914430  51.490690   
Longitude             660653.0  -1.431210  1.383330  -7.516225  -2.332291   
Number_of_Casualties  660679.0   1.357040  0.824847   1.000000   1.000000   
Number_of_Vehicles    660679.0   1.831255  0.715269   1.000000   1.000000   

                            50%        75%        max  
Latitude              52.315641  53.453452  60.757544  
Longitude             -1.411667  -0.232869   1.762010  
Number_of_Casualties   1.000000   1.000000  68.000000  
Number_of_Vehicles     2.000000   2.000000  32.000000

<h2>5. Call Each Column Individually</h2>

In [6]:
accidents['Index']

0         200701BS64157
1         200701BS65737
2         200701BS66127
3         200701BS66128
4         200701BS66837
              ...      
660674    201091NM01760
660675    201091NM01881
660676    201091NM01935
660677    201091NM01964
660678    201091NM02142
Name: Index, Length: 660679, dtype: object

In [7]:
accidents['Accident_Severity']

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [8]:
accidents['Accident Date']

0           5/6/2019
1           2/7/2019
2         26-08-2019
3         16-08-2019
4           3/9/2019
             ...    
660674    18-02-2022
660675    21-02-2022
660676    23-02-2022
660677    23-02-2022
660678    28-02-2022
Name: Accident Date, Length: 660679, dtype: object

In [9]:
accidents['Latitude']

0         51.506187
1         51.495029
2         51.517715
3         51.495478
4         51.488576
            ...    
660674    57.374005
660675    57.232273
660676    57.585044
660677    57.214898
660678    57.575210
Name: Latitude, Length: 660679, dtype: float64

In [10]:
accidents['Light_Conditions']

0               Darkness - lights lit
1                            Daylight
2         Darkness - lighting unknown
3                            Daylight
4               Darkness - lights lit
                     ...             
660674                       Daylight
660675         Darkness - no lighting
660676                       Daylight
660677         Darkness - no lighting
660678                       Daylight
Name: Light_Conditions, Length: 660679, dtype: object

In [11]:
accidents['District Area']

0         Kensington and Chelsea
1         Kensington and Chelsea
2         Kensington and Chelsea
3         Kensington and Chelsea
4         Kensington and Chelsea
                   ...          
660674                  Highland
660675                  Highland
660676                  Highland
660677                  Highland
660678                  Highland
Name: District Area, Length: 660679, dtype: object

In [12]:
accidents['Longitude']

0        -0.209082
1        -0.173647
2        -0.210215
3        -0.202731
4        -0.192487
            ...   
660674   -3.467828
660675   -3.809281
660676   -3.862727
660677   -3.823997
660678   -3.895673
Name: Longitude, Length: 660679, dtype: float64

In [13]:
accidents['Number_of_Casualties']

0         1
1         1
2         1
3         1
4         1
         ..
660674    2
660675    1
660676    1
660677    1
660678    1
Name: Number_of_Casualties, Length: 660679, dtype: int64

In [14]:
accidents['Number_of_Vehicles']

0         2
1         2
2         3
3         4
4         2
         ..
660674    1
660675    1
660676    3
660677    2
660678    1
Name: Number_of_Vehicles, Length: 660679, dtype: int64

In [15]:
accidents['Road_Surface_Conditions']

0                  Dry
1          Wet or damp
2                  Dry
3                  Dry
4                  Dry
              ...     
660674             Dry
660675    Frost or ice
660676    Frost or ice
660677     Wet or damp
660678     Wet or damp
Name: Road_Surface_Conditions, Length: 660679, dtype: object

In [16]:
accidents['Road_Type']

0         Single carriageway
1         Single carriageway
2                        NaN
3         Single carriageway
4                        NaN
                 ...        
660674    Single carriageway
660675    Single carriageway
660676    Single carriageway
660677    Single carriageway
660678      Dual carriageway
Name: Road_Type, Length: 660679, dtype: object

In [17]:
accidents['Urban_or_Rural_Area']

0         Urban
1         Urban
2         Urban
3         Urban
4         Urban
          ...  
660674    Rural
660675    Rural
660676    Rural
660677    Rural
660678    Rural
Name: Urban_or_Rural_Area, Length: 660679, dtype: object

In [18]:
accidents['Weather_Conditions']

0            Fine no high winds
1         Raining no high winds
2                           NaN
3            Fine no high winds
4                           NaN
                  ...          
660674       Fine no high winds
660675       Fine no high winds
660676       Fine no high winds
660677       Fine no high winds
660678    Snowing no high winds
Name: Weather_Conditions, Length: 660679, dtype: object

In [19]:
accidents['Vehicle_Type']

0                                          Car
1                                          Car
2                        Taxi/Private hire car
3         Bus or coach (17 or more pass seats)
4                                Other vehicle
                          ...                 
660674                                     Car
660675                                     Car
660676                                     Car
660677                   Motorcycle over 500cc
660678                                     Car
Name: Vehicle_Type, Length: 660679, dtype: object

<h2>6. Check For Null Values</h2>

In [20]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h2>7. Fill Null Values with statistical values</h2>

In [21]:
accidents['Accident_Severity'] = accidents['Accident_Severity'].fillna(accidents['Accident_Severity'].mode()[0])
accidents['Light_Conditions'] = accidents['Light_Conditions'].fillna(accidents['Light_Conditions'].mode()[0])
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].fillna(accidents['Road_Surface_Conditions'].mode()[0])
accidents['Number_of_Casualties'] = accidents['Number_of_Casualties'].fillna(accidents['Number_of_Casualties'].median())
accidents['Number_of_Vehicles'] = accidents['Number_of_Vehicles'].fillna(accidents['Number_of_Vehicles'].median())
accidents['District Area'] = accidents['District Area'].fillna(accidents['District Area'].mode()[0])

In [22]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions        0
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h2>8. Check Column Data Types</h2>

In [23]:
accidents.info

<bound method DataFrame.info of                 Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1  

## QUESTION'S,CODE, AND INSIGHT'S 

## 1.How many accidents occurred in urban areas compared to rural areas?


In [38]:
accidents['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421663
Rural          238990
Unallocated        11
Name: count, dtype: int64

### Insight: There were 421,663 accidents in urban areas and 238,990 in rural areas.

## 2.What is the average number of casualties per accident?

In [36]:
accidents['Number_of_Casualties'].mean()

np.float64(1.357040257068864)

### Insight: On average, there are 1.35 casualties per accident.

## 3.How many total accidents are recorded?

In [39]:
len(accidents)

660679

## Insight: There are 660,679 accidents in the dataset.

## 4.How many accidents occurred on single carriageways?

In [42]:
(accidents['Road_Type'] == 'Single carriageway').sum()

np.int64(492143)

## Insight: There were 492,143 accidents on single carriageways.

## 5.How many different weather conditions are recorded?

In [43]:
accidents['Weather_Conditions'].nunique()

8

### Insight: There are 8 different weather conditions in the dataset.

## 6.Which weather condition had the most accidents?

In [44]:
accidents['Weather_Conditions'].mode()[0]

'Fine no high winds'

### Insight: The most common weather condition is Fine no high winds.

## 7.Which day had the single highest number of accidents?

In [48]:
accidents['Accident Date'].value_counts().idxmax()

'30-11-2019'

### Insight: The day with the most accidents was 30-11-2019.

## 8.Are accidents more severe in rural or urban areas?

In [55]:
accidents.groupby('Urban_or_Rural_Area')['Accident_Severity'].value_counts()

Urban_or_Rural_Area  Accident_Severity
Rural                Slight               196077
                     Serious               37312
                     Fatal                  5601
Unallocated          Slight                   10
                     Serious                   1
Urban                Slight               367701
                     Serious               50902
                     Fatal                  3060
Name: count, dtype: int64

### Insight: Rural accidents are more often severe/fatal, while urban accidents are more frequent but less severe.

## 9.Is there a correlation between vehicle type and accident severity?

In [59]:
pd.get_dummies(accidents['Vehicle_Type']).corrwith(pd.get_dummies(accidents['Accident_Severity'])['Serious'])

Agricultural vehicle                     0.001808
Bus or coach (17 or more pass seats)    -0.001889
Car                                     -0.000344
Data missing or out of range            -0.001183
Goods 7.5 tonnes mgw and over            0.000281
Goods over 3.5t. and under 7.5t          0.002003
Minibus (8 - 16 passenger seats)         0.000991
Motorcycle 125cc and under              -0.000231
Motorcycle 50cc and under               -0.000050
Motorcycle over 125cc and up to 500cc   -0.000344
Motorcycle over 500cc                    0.000718
Other vehicle                            0.000693
Pedal cycle                              0.003272
Ridden horse                            -0.000966
Taxi/Private hire car                   -0.000129
Van / Goods 3.5 tonnes mgw or under     -0.000145
dtype: float64

### Insight: Indicates whether certain vehicles like buses, trucks are more linked to serious accidents.

## 10.What is the maximum number of casualties in a single accident?

In [61]:
accidents['Number_of_Casualties'].max()

np.int64(68)

### insight: The maximum casualties recorded in a single accident is 68.

## 11.How does weather affect the average number of vehicles per accident?

In [64]:
accidents.groupby('Weather_Conditions')['Number_of_Vehicles'].mean()

Weather_Conditions
Fine + high winds        1.779401
Fine no high winds       1.836809
Fog or mist              1.807823
Other                    1.777668
Raining + high winds     1.777951
Raining no high winds    1.831610
Snowing + high winds     1.737853
Snowing no high winds    1.754889
Name: Number_of_Vehicles, dtype: float64

### Insight: Poor weather like rain, fog often involves slightly more vehicles in each accident.

## 12.Are motorcycles more involved in fatal accidents compared to cars?

In [69]:
accidents.groupby('Vehicle_Type')['Accident_Severity'].value_counts()

Vehicle_Type                           Accident_Severity
Agricultural vehicle                   Slight                 1644
                                       Serious                 282
                                       Fatal                    21
Bus or coach (17 or more pass seats)   Slight                22180
                                       Serious                3373
                                       Fatal                   325
Car                                    Slight               424954
                                       Serious               66461
                                       Fatal                  6577
Data missing or out of range           Slight                    6
Goods 7.5 tonnes mgw and over          Slight                14770
                                       Serious                2321
                                       Fatal                   216
Goods over 3.5t. and under 7.5t        Slight                 5172
     

### Insight: Motorcycles have a higher proportion of fatal crashes than cars highlighting their vulnerability.

## 13.How does day/night affect the number of vehicles involved?

In [74]:
accidents.groupby('Light_Conditions')['Number_of_Vehicles'].mean()

Light_Conditions
Darkness - lighting unknown    1.758482
Darkness - lights lit          1.775498
Darkness - lights unlit        1.752654
Darkness - no lighting         1.599781
Daylight                       1.865385
Name: Number_of_Vehicles, dtype: float64

### Insight: Daytime accidents usually involve more vehicles, while nighttime accidents often involve fewer but are more severe.

## 14.Do accidents in rural areas have higher fatality rates at night?

In [76]:
accidents.groupby(['Urban_or_Rural_Area','Light_Conditions'])['Accident_Severity'].value_counts()

Urban_or_Rural_Area  Light_Conditions             Accident_Severity
Rural                Darkness - lighting unknown  Slight                 2082
                                                  Serious                 344
                                                  Fatal                    41
                     Darkness - lights lit        Slight                20290
                                                  Serious                3827
                                                  Fatal                   578
                     Darkness - lights unlit      Slight                  782
                                                  Serious                 150
                                                  Fatal                    29
                     Darkness - no lighting       Slight                27087
                                                  Serious                6861
                                                  Fatal                  1

### Insight: Fatal accidents are most likely in rural areas during darkness without street lights.

## 15.How do lighting conditions affect the number of casualties per accident?

In [78]:
accidents.groupby('Light_Conditions')['Number_of_Casualties'].mean()

Light_Conditions
Darkness - lighting unknown    1.306447
Darkness - lights lit          1.369374
Darkness - lights unlit        1.331105
Darkness - no lighting         1.541657
Daylight                       1.340309
Name: Number_of_Casualties, dtype: float64

### Insight: Poor lighting increases the severity of accidents and raises casualties.

## 16.What percentage of accidents occurred in fine weather?

In [83]:
(accidents['Weather_Conditions'].value_counts(normalize=True)['Fine no high winds'] * 100).round(2)

np.float64(80.56)

### Insight: About 85.56% of accidents occurred in fine weather.

## 17.How many accidents occurred on wet roads?

In [85]:
(accidents['Road_Surface_Conditions'] == 'Wet').sum()

np.int64(0)

### Insight: There were 0 accident on wet roads.

## 18.Which accident severity is most common?

In [86]:
accidents['Accident_Severity'].mode()[0]

'Slight'

### Insight: The most common severity is Slight accidents.

## 19.How many accidents happened in daylight?

In [89]:
(accidents['Light_Conditions'] == 'Daylight').sum()

np.int64(484880)

### Insight: There were 484,880 accidents in daylight.

## 20.How many accidents involved more than 10 vehicles?

In [91]:
(accidents['Number_of_Vehicles'] > 10).sum()

np.int64(42)

### Insight: There were 42 accidents involving more than 10 vehicles.